cpp/src/fil/common.cuh

```diff
@@ -38,81 +38,92 @@ __host__ __device__ __forceinline__ int forest_num_nodes(int num_trees, int dept
   return num_trees * tree_num_nodes(depth);
 }
 
+template <typename F>
 struct storage_base {
   categorical_sets sets_;
-  float* vector_leaf_;
+  F* vector_leaf_;
   bool cats_present() const { return sets_.cats_present(); }
 };
 
-/** dense_tree represents a dense tree */
-
-struct dense_tree : tree_base {
-  __host__ __device__ dense_tree(categorical_sets cat_sets, dense_node* nodes, int node_pitch)
+/** represents a dense tree */
+template <typename F>
+struct tree<dense_node<F>> : tree_base {
```

Comment:
Is it OK to define a specialization before the template type definition?

Reply:
Apparently, template specializations can appear after the type _declaration_; no definition is required, see e.g.: 

https://en.cppreference.com/w/cpp/language/template_specialization
https://en.cppreference.com/w/cpp/language/partial_specialization

So this is resolved.